### PR TITLE
add start to RangeDecoder

### PR DESCRIPTION
**Pull request by hiroki-kojima**
I propose an implementation of `start` that can be executed at any time instead of removing the RangeDecoder's constructor.

The motivation for the suggestion is that the `bytes` may not be ready before the constructor is called.

### CODE REVIEW

**Comment by diegodox** [comment]
With this PR, `RangeDecoder.start(data)` can be called multiple time, even if `RangeDecoder` is already started. Moreover, `RangeDecode.decode()` can be called without any data given.
We can note "Don't call this method twice or more. That's not make any sense." on `start` and "Confirm you have called `start`, before calling this method" on `decode`.

I want that whenever `RangeDecoder` is constructed, it is ready to decode.

You can delay construction with `std::optional<RangeDecoder>`.
Reference(jp): https://cpprefjp.github.io/reference/optional/optional.html

**Comment by hiroki-kojima** [comment]
@diegodox 
ok, totally agree with you.

**Comment by diegodox** [comment]
Since re-discussion is requested, reopen this.
https://github.com/diegodox/rangecoder-cpp17/issues/7#issuecomment-999373834

**Comment by diegodox** [comment]
RangeDecoder must responsible for consistency between internal data and range. ( I mean it's not user's respoinsible, User only have responsiblity to consistency between data and pmodel. )

I'm welcome supporting delay initialization, but it's must keep consistency between internal data and range.

**Comment by diegodox** [comment]
Workaround: You can wrap `RangeCoder`. ( little overhead but may be acceptable)

```cpp
class Deocder {
    public:
        Decoder() { m_range_decoder = rangecoder::RangeDecoder(std::queue()); }
        initialize(std::queue<rangecoder::byte_t> data) { m_range_decoder = rangecoder::RangeDecoder(data); }
        decode(Pmodel pmodel, int index) { .. }
    private:
        rangecoder::RangeDecoder m_range_decoder,
        std::vector<int> m_decoded;
}
```

File: rangecoder.h [hunk 1]
```diff
@@ -217,11 +217,9 @@ namespace rangecoder
     class RangeDecoder : RangeCoder
     {
     public:
-        RangeDecoder() = delete;
-
-        RangeDecoder(std::queue<byte_t> bytes)
+        void start(std::queue<byte_t> bytes)
```
Comment (diegodox) [inline rangecoder.h:220]: rename to initialize
```suggestion
        void initialize(std::queue<byte_t> bytes)
```

File: rangecoder.h [hunk 2]
```diff
@@ -217,11 +217,9 @@ namespace rangecoder
...
         {
-            m_bytes = bytes;
+            m_bytes = std::move(bytes);
```
Comment (diegodox) [inline rangecoder.h:223]: initialize range here
```suggestion
        m_lower_bound = 0;
        m_range = std::numeric_limits<range_t>::max();
```

**Review by diegodox — CHANGES_REQUESTED** [review]
rename to initialize and add initialization of range to keep consistency.

**Comment by hiroki-kojima** [comment]
@diegodox 
thank you for your consideration!

**Comment by diegodox** [comment]
Apart from previous comment, Adding some methods to add data after initialization ( that method may or may be named `append_data` that take bytes as argument and push to internal bytes queue `m_bytes`) is in my mind.

File: rangecoder.h (hunk 2 above)
Comment (diegodox) [reply rangecoder.h:223]: mark unresolved again, I think this is not resolved. Please give me reason you don't initialize range here.

**Comment by diegodox** [comment]
> Apart from previous comment, Adding some methods to add data after initialization ( that method may be named `append_data` that take bytes as argument and push to internal bytes queue `m_bytes`) is in my mind.

Just a note: This will be another issue.

**Comment by diegodox** [comment]
@hiroki-kojima 
- CI test fails since m_lower_bound and m_range is declared as private in RangeCoder.
Maybe we have two choices:
  1. Declare m_range and m_lower_bound as protected
  2. Create initialize method on RangeCoder
- CI check format fails since indentation broken around added code.

**Comment by hiroki-kojima** [comment]
@diegodox 
i suggest introducing setters for `m_range` and `m_lower_bound`.

https://github.com/diegodox/rangecoder-cpp17/pull/6/commits/9c75c41cb0440f25e29bdb98f9f8615835c1314f

what do you think?

**Comment by hiroki-kojima** [comment]
The antonym for `finish` is `start`.
On the encoder side, `finish` is executed last, so I thought that `start` was preferable to `init` (or `initialize `) as the name of the opposite operation.

**Comment by diegodox** [comment]
> @diegodox 
> i suggest introducing setters for `m_range` and `m_lower_bound`.
> 
> https://github.com/diegodox/rangecoder-cpp17/pull/6/commits/9c75c41cb0440f25e29bdb98f9f8615835c1314f
> 
> what do you think?

What is merit about it?

RangeCoder is not intended to be used by nothing else RangeEncoder and RangeDecoder.

There are no clear pre-condition expression about range and lower_bound.
I'm not in C++ culture, so I don't know whether setter is widly used without checking pre-condition or not.

As Rust user, seems weird.

**Comment by hiroki-kojima** [comment]
The class only needs to know the interface (in public and protected) of the base class.
With encapsulation, derived classes don't have to worry about the existence of prerequisites in the first place.
> There are no clear pre-condition expression about range and lower_bound.

**Comment by hiroki-kojima** [comment]
Getters were already implemented. Therefore, I think the addition of setters is not unnatural.

**Comment by diegodox** [comment]
@hiroki-kojima

>The class only needs to know the interface (in public and protected) of the base class.
>With encapsulation, derived classes don't have to worry about the existence of prerequisites in the first place.
As I mentioned before, RangeCoder is not intended to be used by nothing else RangeEncoder and RangeDecoder.
So I didn't find advantage in this situation.

But, protected member or protected setter is not important problem, I'm not serious about it. (At least simple and readble enough for me.)
We can go with protected setter 😄 .

**Review by diegodox — APPROVED** [review]
LGTM

**Pull request merged**